rstudio/rsconnect
# Disallow deploying `.qmd` files with `rmd-*` app modes

### Intent

Previously, Quarto documents and projects would deploy as `rmd-static` or `rmd-shiny` if they were deployed without Quarto info available (either via running `quarto inspect` or via the `metadata` argument).

With this PR, `rsconnect` will stop with an error if that condition occurs, and advise the user to pass the path to a Quarto binary to the `quarto` argument.

Fixes #594 

### Approach

Previously, we would infer one set of `rmdFiles` that would contain both Quarto and R Markdown files, and decide between `rmd-*` and `quarto-*` app modes based on the presence or absence of `quartoInfo`, which is the inferred Quarto info from either running Quarto Inspect or unpacking metadata.

Now, we find `.Rmd` and `.qmd` files separately. The list of file names is stored in `rmdFiles` and `qmdFiles` separately, and a logical vector of whether they're `runtime: shiny` is stored in `shinyRmdFiles` and `shinyQmdFiles`. We also look for `^_quarto.y(a)?ml$`.

A deployment requires Quarto if… I was going to write out the condition, but the code is more readable than English.

```r
  hasQuartoSupportedFiles <- any(length(qmdFiles) > 0, length(rmdFiles > 0))
  requiresQuarto <- (hasQuartoSupportedFiles && hasQuartoYaml) || length(qmdFiles) > 0
```

*Any* `.qmd` files make us require Quarto info. or, if we have Quarto-compatible files and a `_quarto.yml`, we need Quarto info. In those cases, if we don't have it, we raise an error.

After this point, we take the presence of the `quartoInfo` variable to determine whether we use `quarto` or `rmd` app modes, in each locations where we assign `*-shiny` and `*-static`. This still will let you deploy R Markdown files as Quarto content, but prevents you from deploying Quarto content as non-Quarto.

The RStudio IDE was adding the line `PackageCleanBeforeInstall: Yes` to `.Rproj` files for a while. Now it apparently removes that line. I committed the change, because it happened with every install of a package.

### Automated Tests

Three tests were added:
- Attempt to create a manifest for `quarto-website-r`, leaving the `quarto` argument `NULL`, and expect the error message.
- Attempt to create a manifest for `quarto-doc-none`, leaving the `quarto` argument `NULL`, and expect the error message.
- Attempt to create a manifest for `test-reds/simple.Rmd`, passing in `quarto`. Ensure that the manifest has the `quarto-static` app mode and `knitr` engine.

### QA Notes

Install this branch of `rsconnect` from the RStudio IDE.

Quarto content in the `tests/testthat` directory should deploy correctly:

```r
rsconnect::deployApp(appDir = "tests/testthat/quarto-website-r", server = "localhost", quarto = quarto::quarto_path())
```

Without passing the `quarto` argument, you should see an error:

```r
rsconnect::deployApp(appDir = "tests/testthat/quarto-website-r", server = "localhost")
```

After testing, you'll want to go into the `tests/testthat/quarto-website-r` folder and remove the `rsconnect` folder that was created.

Deploying an R Markdown document from the IDE with this version installed should still deploy with an R Markdown app mode. (I tested this myself, but it's good to verify in a different environment.)

### Checklist
